[Shared] Config and genesis handling refactoring idea

deblasis opened the pull request:
## Description

This PR is a way to convey a design idea for the refactoring of the way we handle `config` and `genesis` as highlighted in a couple of Protocol Hours and also summarized, along with more context that should help us making the right decisions, into the discussion https://github.com/pokt-network/pocket/discussions/234

## Issue

There's no issue per-se, this is basically TECHDEBT handling I guess. Let me know if I should create one.

## Type of change

Please mark the options that are relevant.

- [x] New feature, functionality or library
- [x] Code health or cleanup

## List of changes

- Abstracted config and genesis handling
- Mockable runtime
- Refactored all modules to use RuntimeMgr
- Updated RuntimeMgr to manage clock as well
- Modules now accept `interfaces` instead of paths.
- Unmarshalling is done in a new `runtime` package (runtime because what we do in there affects the runtime of the application)
- We are now able to accept configuration via environment variables (thanks to @okdas for inspiration and [sp13 for Viper]("github.com/spf13/viper"))


## Testing

- [x] `make test_all`
- [x] [LocalNet](https://github.com/pokt-network/pocket/blob/main/docs/development/README.md) w/ all of the steps outlined in the `README`

## Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have tested my changes using the available tooling
- [x] If applicable, I have made corresponding changes to related local or global README
- [x] If applicable, I have added new diagrams using [mermaid.js](https://mermaid-js.github.io)
- [x] If applicable, I have added tests that prove my fix is effective or that my feature works

deblasis committed:
refactor(Config): runtime module

deblasis committed:
style(Client): consistency

deblasis committed:
fix(Config): make configPath and genesisPath available in Base

deblasis committed:
style(Persistence): Persistance -> Persistence

deblasis committed:
fix(Config): Base init

deblasis committed:
feat(Config): viper config parsing + environment variables

deblasis committed:
fix(Config): filename handling

andrewnguyen22 commented on p2p/module_raintree_test.go:
```diff
@@ -281,7 +283,8 @@ func generateKeys(_ *testing.T, numValidators int) []cryptoPocket.PrivateKey {
 // A mock of the application specific to know if a message was sent to be handled by the application
 // INVESTIGATE(olshansky): Double check that how the expected calls are counted is accurate per the
-//                         expectation with RainTree by comparing with Telemetry after updating specs.
+//
```
Getting some failed tests:

```
module_raintree_test.go:386
        	            				module_raintree_test.go:209
        	            				module_raintree_test.go:180
        	Error:      	Received unexpected error:
        	            	Config File "config" Not Found in "[/Users/andrewnguyen/go/src/github.com/pokt-network/pocket/p2p]"
        	Test:       	TestRainTreeCompleteTwentySevenNodes
```

andrewnguyen22 replied:
Awesome :)

andrewnguyen22 commented on runtime/config.go:
```diff
@@ -0,0 +1,89 @@
+package runtime
```
My general issue with this package as described in https://github.com/pokt-network/pocket/discussions/234

Is that we essentially allow for other modules to access `module specific` configuration and genesis states.

This is not at all the intention

This seems like an error prone architecture to me

deblasis replied:
I see what you mean but if we think about configurability in general isn't the environment (and therefore environment variables, configuration files, etc) something that's accessible everywhere within the app one way or another? 

What we should guard against is not the mere access but whenever for example inside the P2P module I use the PostgresUrl (part of persistence config) to build some logic on top of. Obviously, that's fragile and spaghetti code and bug prone because someone might just change that in persistence and consequently break P2P for example.

The environment is a `os.GetEnv()` call away and therefore any config value as well potentially.
The point is: will it make it to `main`?
Same applies for accessing configuration/genesis files.

If we want the modules to be completely isolated from each other and to bring to zero the shared code I think they should be running as separate binaries on their own isolation context (Docker container or whatnot) and communicate with the other modules via some form of well defined RPC. But we are still in monolith land here and everything is a tradeoff. Anyway, I am glad I can pick your brain about this kind of stuff. 🙏

andrewnguyen22 commented on shared/modules/module.go:
```diff
@@ -18,9 +19,3 @@ type InterruptableModule interface {
 	Stop() error
 }
-
-type InitializableModule interface {
```
When we talk about `code clarity` ; here's another example where we loosen interfaces. It's difficult for me to say that the `V1 - Structs` design is definitely `clearer`

andrewnguyen22 commented on shared/node.go:
```diff
@@ -28,32 +29,38 @@ type Node struct {
 	}
 
-	p2pMod, err := p2p.Create(configPath, genesisPath, false)
+	p2pMod, err := p2p.Create(cfg.P2P, false)
```
The design pattern for module creation is loosened here

deblasis replied:
Maybe this is better
![image](https://user-images.githubusercontent.com/29378614/191291630-bf413a2f-1e86-48cb-8a6a-0d1a3303e1ad.png)

maybe builder is not the best naming here

![image](https://user-images.githubusercontent.com/29378614/191291874-93df89ed-8686-4ae5-afa2-c6846a8cf981.png)

![image](https://user-images.githubusercontent.com/29378614/191291920-31cb9ba5-8188-47a4-8cae-5e47c5dcb9ad.png)

![image](https://user-images.githubusercontent.com/29378614/191292206-455e3163-ae51-4c77-bf1f-d13f6d64057c.png)

andrewnguyen22 replied:
Yes, the `getRandomPK` should go away soon, but I like this a log better if you guys do

++ I renaming NewBuilder though

andrewnguyen22 commented on shared/test_artifacts/generator.go:
```diff
@@ -34,64 +41,64 @@ var (
-	vals, validatorPrivateKeys := NewActors(MockActorType_Val, numValidators)
-	serviceNodes, snPrivateKeys := NewActors(MockActorType_Node, numServiceNodes)
-	fish, fishPrivateKeys := NewActors(MockActorType_Fish, numFisherman)
+	apps, appsPrivateKeys := NewActors(types.UtilActorType_App, numApplications)
```
I like that we deleted the mocks!

I don't think this change is exclusive to either `v1 strings or structs` implementation however

deblasis replied:
I tried to cater to both point of views and hacked some code to see if something in between could be feasible of not.

andrewnguyen22 commented on utility/test/module_test.go:
```diff
@@ -8,6 +8,7 @@ import (
 	"os"
 	"testing"
 
+	typesPers "github.com/pokt-network/pocket/persistence/types"
```
`Utility mod` added import from `Persistence mod`. 

We're trying to get rid of this cross module sharing

deblasis replied:
Perhaps we could move module creation for testing purposes under test_artifacts to hide away the implementation details of the other modules when we are testing one

Olshansk replied:
Created https://github.com/pokt-network/pocket/issues/239

okdas replied:
@Olshansk that is likely going to be a somewhat sophisticated AST rule but should be doable! 👍

andrewnguyen22 replied:
I'm not sure how this was resolved. We just added an issue to automatically check

deblasis committed:
refactor(Config): runtime module and InitializableModule interface

deblasis committed:
feat(Config): useRandomPK in runtime

deblasis committed:
refactor(Config): renamed Builder / updated signatures

deblasis committed:
style(Config): builder -> runtime

deblasis committed:
feat(Config): implement InitlializableModule in all modules

deblasis committed:
fix(Consensus): remove pacemakerConfig from interface

deblasis committed:
feat(Config): params/flags injection (stab @private key for CLI)

Olshansk reviewed (changes requested):
I think my review is a bit outdated but figured I'd publish the comments I have as an interim iteration. Looking great so far

deblasis committed:
style(Client): cleanup

deblasis commented on app/client/main.go:
```diff
@@ -47,24 +48,32 @@ var consensusMod modules.ConsensusModule
+	cfg := runtime.GetConfig()
+	genesis := runtime.GetGenesis()
+
+	cons, err := consensus.Create(runtime) // TECHDEBT: extra param required for injecting private key hack for debug client
```
Addressed in 2fec88b

deblasis commented on app/client/main.go:
```diff
@@ -47,24 +48,32 @@ var consensusMod modules.ConsensusModule
 		log.Fatalf("[ERROR] Failed to create consensus module: %v", err.Error())
 	}
-	p2pMod, err = p2p.Create(defaultConfigPath, defaultGenesisPath, true) // TECHDEBT: extra param required for injecting private key hack for debug client
+	p2pm, err := p2p.Create(runtime) // TECHDEBT: extra param required for injecting private key hack for debug client
```
Addressed in 2fec88b

deblasis commented on app/client/main.go:
```diff
@@ -47,24 +48,32 @@ var consensusMod modules.ConsensusModule
 	// code paths are executed. To avoid those messages interfering with the telemetry data collected, a non-nil telemetry
 	// module that NOOPs (per the configs above) is injected.
-	telemetryMod, err := telemetry.Create(defaultConfigPath, defaultGenesisPath)
+	telemetryM, err := telemetry.Create(runtime)
```
Addressed in 2fec88b

deblasis committed:
feat(Shared): ConfigurableModule and GenesisDependentModule 👀

deblasis committed:
Merge remote-tracking branch 'upstream/main' into issue/config_genesis_refactoring1

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>

deblasis committed:
chore(go.mod): tidy

deblasis commented on consensus/module.go:
```diff
@@ -65,35 +64,35 @@ type ConsensusModule struct {
+	cfg := runtime.GetConfig()
+	genesis := runtime.GetGenesis()
+
+	moduleCfg := cfg.Consensus.(*typesCons.ConsensusConfig)
```
The problem is that when we return `any` we would still have to cast to the module specific config. Validation is something that we can add a signature for but casting unfortunately has to be done on the calling site I think.

I took a stab at the concept, please have a look at https://github.com/pokt-network/pocket/pull/235/commits/25979e351e0da2f5b43ead67245afa97e773c9be

If you likey I am gonna apply the same concept to all the other modules as well, also what's nice is that we would know if a module needs genesis or config or both with these interfaces. Naming is hard so suggestions are welcome :)

deblasis replied:
Addressed in https://github.com/pokt-network/pocket/pull/235/commits/f5ea7e47c9d7c1b2e84b08cb40a0ca70c61299b5

deblasis committed:
feat(Shared): KeyholderModule 👀

deblasis commented on consensus/module.go:
```diff
@@ -65,35 +64,35 @@ type ConsensusModule struct {
+	valMap := typesCons.ValidatorListToMap(moduleGenesis.Validators)
 	var privateKey cryptoPocket.PrivateKey
-	if useRandomPK {
+	if runtime.ShouldUseRandomPK() {
```
Have a look at https://github.com/pokt-network/pocket/pull/235/commits/7c5b94399f48b7587336b621274edc41c20c406a, if you like the concept I'll apply elsewhere too. 

We could argue that this line https://github.com/pokt-network/pocket/commit/7c5b94399f48b7587336b621274edc41c20c406a#diff-eba95a8b904f8186683d2f9c106a3dc021f014918420e6a074d0a2b4e59c261dR210

could sound a bit "cryptic", pun intended, but we are good citizens and config has been already validated when we are trying to extract stuff from it, eg: `PrivateKey`

Olshansk replied:
> could sound a bit "cryptic"

( ͡° ͜ʖ ͡°)

In general, I don't **love** this solution but I think it's more than good enough to get us going.

deblasis committed:
fix(Shared): Enforcing  for all modules

deblasis commented on runtime/runtime.go:
```diff
@@ -0,0 +1,108 @@
+}
+
+func New(configPath, genesisPath string, options ...func(*RuntimeConfig)) *RuntimeConfig {
+	b := &RuntimeConfig{
```
the struct used to be called Builder in its infancy but now it's a teen RuntimeConfig even if some folks still call it "b" :) 

Enough, time to grow up!

deblasis replied:
Addressed in https://github.com/pokt-network/pocket/pull/235/commits/69af7f16829f0be9f49c83b6ea547cc09937b5fd

deblasis commented on runtime/runtime.go:
```diff
@@ -0,0 +1,108 @@
+	b.genesis = genesis
+
+	for _, o := range options {
+		o(b)
```
Correct, I would say, and correct me if I am wrong: this allow for flexibility, can it be improved? Maybe. Can this be used maliciously, of course. 

In Go you are the owner of the code and you can do anything with it. In other languages you can enforce immutability and there are well-known patterns, in Go I have seen something like wrapping values in a struct but there are walkarounds and it makes the code hard to understand IMHO. 

Open to suggestions here, happy to learn something new

Olshansk replied:
I do like this approach and was simply confirming. +💯

deblasis committed:
refactor(Runtime): renamed b to rc

deblasis commented on runtime/runtime.go:
```diff
@@ -0,0 +1,108 @@
+	viper.AutomaticEnv()
+
+	err = viper.ReadInConfig()
+	if err != nil {
```
Addressed in https://github.com/pokt-network/pocket/pull/235/commits/69af7f16829f0be9f49c83b6ea547cc09937b5fd

deblasis commented on runtime/runtime.go:
```diff
@@ -0,0 +1,108 @@
+	config.Base.ConfigPath = b.configPath
+	config.Base.GenesisPath = b.genesisPath
+
+	genesis, err = ParseGenesisJSON(b.genesisPath)
```
Yes, currently accessed by persistence in HandleDebugMessage
https://github.com/deblasis/pocket/blob/69af7f16829f0be9f49c83b6ea547cc09937b5fd/persistence/debug.go#L18-L19

deblasis commented on shared/modules/runtime_module.go:
```diff
@@ -0,0 +1,7 @@
+type Runtime interface {
+	GetConfig() Config
+	GetGenesis() GenesisState
+	ShouldUseRandomPK() bool
```
I put it there exactly because there's a semantic conflict between that flag and the privateKey in config.

If this was only used in tests, yeah, we can mock that but currently the client needs a random private key to work for debugging purposes. I concur, something to think about.

deblasis commented on persistence/test/setup_test.go:
```diff
@@ -107,11 +108,14 @@ func newTestPersistenceModule(databaseUrl string) modules.PersistenceModule {
 	createTestingGenesisAndConfigFiles(cfg, genesisState)
-	persistenceMod, err := persistence.Create(testingConfigFilePath, testingGenesisFilePath)
+
+	runtime := runtime.New(testingConfigFilePath, testingGenesisFilePath)
```
Nothing stops us from having a `runtime.NewFromSomething(something Something)` 

If you can provide practical examples / an issue, we can totally do that

Olshansk replied:
Resolving!

deblasis committed:
style(Shared): renamed moduleCfg to moduleNameCfg

deblasis commented on p2p/module.go:
```diff
@@ -42,55 +42,39 @@ func (m *p2pModule) GetAddress() (cryptoPocket.Address, error) {
-	l, err := CreateListener(cfg)
+
+	cfg := runtime.GetConfig()
+	moduleCfg := cfg.P2P.(*typesP2P.P2PConfig)
```
Addressed in https://github.com/pokt-network/pocket/pull/235/commits/56be6750b57fa284eb86f92e6b58445faa2738fc for all modules that used moduleCfg

deblasis commented on p2p/module.go:
```diff
@@ -22,6 +20,8 @@ import (
 
 var _ modules.P2PModule = &p2pModule{}
+var _ modules.Module = &p2pModule{}
+var _ modules.InitializableModule = &p2pModule{}
```
Addressed in https://github.com/pokt-network/pocket/pull/235/commits/49980dc8a92dbe8a7f80302f9ce2e130001ed8ac

deblasis committed:
Merge branch 'main' into issue/config_genesis_refactoring1

Olshansk commented on shared/modules/runtime_module.go:
```diff
@@ -0,0 +1,7 @@
+type Runtime interface {
+	GetConfig() Config
+	GetGenesis() GenesisState
+	ShouldUseRandomPK() bool
```
```suggestion
	ShouldUseRandomPK() bool // INVESTIGATE: look into how we can remove this from the runtime interface
```

Olshansk commented on runtime/config.go:
```diff
@@ -0,0 +1,106 @@
+	return c.P2P.ConsensusPort
+}
+
+func (c *Config) GetUseRainTree() bool {
```
Similar to how you hid the pacemaker config from here, can you also do the same thing for RainTree?

I think one can argue that if someone implements a `kademila` or `gemini` P2P layer, it shouldn't know of `RainTree`.

deblasis replied:
Addressed in https://github.com/pokt-network/pocket/pull/235/commits/dd78b2ead26dc5ec6373e412a75e5ee1837a132e

Olshansk reviewed (changes requested):
I think one more round of reviews and we'll get over the 🏁

deblasis committed:
fix(SHARED): build error for importing unnecessary package

deblasis committed:
Update shared/modules/runtime_module.go

Co-authored-by: Daniel Olshansky <olshansky.daniel@gmail.com>

deblasis committed:
refactor(runtime): renamed RuntimeConfig -> runtimeConfig

deblasis committed:
fix(runtime): improved WithRandomPK()

deblasis committed:
Merge branch 'issue/config_genesis_refactoring1' of github.com:deblasis/pocket into issue/config_genesis_refactoring1

deblasis commented on runtime/runtime.go:
```diff
@@ -0,0 +1,108 @@
+
+var _ modules.Runtime = &RuntimeConfig{}
+
+type RuntimeConfig struct {
```
Addressed in https://github.com/pokt-network/pocket/pull/235/commits/c3ac3cc55bd9d59f79c1918cb5a88caa24e6dfa9

deblasis commented on runtime/runtime.go:
```diff
@@ -0,0 +1,108 @@
+	config  *Config
+	genesis *Genesis
+
+	useRandomPK bool
```
I think I bypassed the problem via https://github.com/pokt-network/pocket/pull/235/commits/66b4806adb2aebea3fbbedcfa7b067921656b9f4

deblasis committed:
docs(runtime): // RuntimeConfig option helpers comment

deblasis commented on runtime/runtime.go:
```diff
@@ -0,0 +1,108 @@
+	if err != nil {
+		log.Fatalf("[ERROR] Failed to initialize runtime builder: %v", err)
+	}
+	rc.config = cfg
```
I wouldn't put side-effects in `init`, we are accessing the filesystem and we want to have control also on -when- that happens.

deblasis commented on shared/modules/module.go:
```diff
@@ -21,6 +23,17 @@ type InterruptableModule interface {
+	Create(runtime Runtime) (Module, error)
+}
+
+type ConfigurableModule interface {
```
My Ouija board sometimes delivers :)

deblasis committed:
refactor(Shared): nodeModule, P2PAddressableModule

deblasis commented on shared/node.go:
```diff
@@ -27,37 +28,52 @@ type Node struct {
 	Address cryptoPocket.Address
```
Addressed in https://github.com/pokt-network/pocket/pull/235/commits/d663287b23063f9d9d0c701d54087942a0874bb8

deblasis committed:
refactor(Consensus): interface embedding

deblasis commented on consensus/module.go:
```diff
@@ -22,10 +19,17 @@ const (
+	_ modules.ConsensusGenesisState = &typesCons.ConsensusGenesisState{}
+
+	_ modules.PacemakerConfig = &typesCons.PacemakerConfig{}
+	_ modules.ConsensusConfig = &typesCons.ConsensusConfig{}
```
Addressed in https://github.com/pokt-network/pocket/pull/235/commits/9e5602b0fed0c066875072df67c9767d1af515fd

deblasis committed:
refactor(Consensus): consistency in naming

deblasis committed:
refactor(Persistence): consistence in naming

deblasis commented on consensus/module.go:
```diff
@@ -65,47 +69,53 @@ type ConsensusModule struct {
-	cfg := c.(*typesCons.ConsensusConfig)
-	genesis := g.(*typesCons.ConsensusGenesisState)
-	leaderElectionMod, err := leader_election.Create(cfg, genesis)
+	moduleGenesis := genesis.ConsensusGenesisState.(*typesCons.ConsensusGenesisState)
```
Addressed in https://github.com/pokt-network/pocket/pull/235/commits/fb873b6fead51aef65b563791fc41b2c1ace8af5

deblasis commented on persistence/module.go:
```diff
@@ -36,19 +34,19 @@ const (
+	genesis := runtime.GetGenesis()
+
+	persistenceCfg := cfg.Persistence.(*types.PersistenceConfig)
+	moduleGenesis := genesis.PersistenceGenesisState.(*types.PersistenceGenesisState)
```
Addressed in https://github.com/pokt-network/pocket/pull/235/commits/8d26fd389701e23295921b6de5a709722011329b

deblasis committed:
feat(P2P): P2P module now implements ConfigurableModule

deblasis committed:
fix(Persistence): improved DEBUG_CLEAR_STATE

deblasis commented on persistence/debug.go:
```diff
@@ -15,11 +15,11 @@ func (m *PersistenceModule) HandleDebugMessage(debugMessage *debug.DebugMessage)
 	case debug.DebugMessageAction_DEBUG_CLEAR_STATE:
 		m.clearState(debugMessage)
-		g, err := m.InitGenesis(m.genesisPath)
+		g, err := runtime.ParseGenesisJSON(m.GetBus().GetConfig().Base.GenesisPath)
```
Addressed in https://github.com/pokt-network/pocket/pull/235/commits/c6c95dd570d9b9cdb0cf0008af3f8d248d16da5f

deblasis committed:
Merge remote-tracking branch 'upstream/main' into issue/config_genesis_refactoring1

Olshansk commented on consensus/module.go:
```diff
@@ -216,6 +188,24 @@ func (m *ConsensusModule) SetBus(pocketBus modules.Bus) {
+	return nil
+}
+
+func (*ConsensusModule) GetPrivateKey(runtime modules.Runtime) (cryptoPocket.PrivateKey, error) {
```
Do we not need the `sholdUseRandomPk` part here?

deblasis replied:
not anymore

![image](https://user-images.githubusercontent.com/29378614/192514698-6deed7eb-d88d-4121-a9ac-bf7d1e3bba27.png)


😉

Olshansk replied:
Nice!

Olshansk commented on consensus/module.go:
```diff
@@ -22,10 +19,12 @@ const (
-var _ modules.ConsensusConfig = &typesCons.ConsensusConfig{}
-var _ modules.ConsensusModule = &ConsensusModule{}
+var (
+	_ modules.ConsensusModule = &ConsensusModule{}
+
+	_ modules.ConsensusConfig       = &typesCons.ConsensusConfig{}
```
```suggestion
	_ modules.ConsensusModule = &ConsensusModule{}
	_ modules.ConsensusConfig       = &typesCons.ConsensusConfig{}
```

Olshansk commented on p2p/module.go:
```diff
@@ -125,7 +107,7 @@ func (m *p2pModule) Start() error {
 	}
 
-	if m.p2pConfig.GetUseRainTree() {
+	if m.p2pConfig.(*typesP2P.P2PConfig).UseRainTree {
```
I feel like we can locally store the casted type on line 31.

deblasis replied:
Addressed in https://github.com/pokt-network/pocket/pull/235/commits/c60206fc430c0b2a8c93ec473d9626aaebadef20

Also applied the same pattern to the other modules for consistency and also removed config related fields from the modules to use the config instead.

Olshansk reviewed (approved):
Going to call out I'm pretty sure we're all aware we'll keep iterating on this in main over time, but I really like where it's going so LFG!

deblasis committed:
feat(Shared): modules implement ConfigurableModule

deblasis committed:
refactor(Shared): consistency: adding config *structType in modules

deblasis committed:
Update consensus/module.go

Co-authored-by: Daniel Olshansky <olshansky.daniel@gmail.com>

deblasis committed:
fix(Runtime): fix WithPK for when called with a nil P2PConfig

deblasis commented:
> Going to call out I'm pretty sure we're all aware we'll keep iterating on this in main over time, but I really like where it's going so LFG!

I think this needs a nod from @andrewnguyen22 as well

Olshansk commented:
> I think this needs a nod from @andrewnguyen22 as well

Agreed. LFG after we get a 👍 from @andrewnguyen22, or not go yet 😅

andrewnguyen22 commented on app/client/main.go:
```diff
@@ -47,24 +48,32 @@ var consensusMod modules.ConsensusModule
-	consensusMod, err = consensus.Create(defaultConfigPath, defaultGenesisPath, true) // TECHDEBT: extra param required for injecting private key hack for debug client
+
+	runtime := runtime.New(defaultConfigPath, defaultGenesisPath, runtime.WithRandomPK())
+	cfg := runtime.GetConfig()
```
I don't understand why all modules get passed `runtime` object (which contains config and genesis) and yet, the creation of the `Bus` takes the config and genesis objects separately. Please pass the runtime object to the `bus creation`

deblasis replied:
Updated.

Since we removed `GetConfig()` and `GetGenesis()` from the `bus`, I am currently not passing in the `runtimeMgr` in 
https://github.com/pokt-network/pocket/pull/235/files#diff-91e861a0bd67a2298f7a42d2ad6f060e8b59153fab4519a9f6bfd4fa293919c4R75 
and also in the `CreateBus()`.

Please let me know if you think it could be beneficial have a reference to the runtime accessible via the bus. But anyway we can always consider adding it when/if needed

andrewnguyen22 commented on app/client/main.go:
```diff
@@ -47,24 +48,32 @@ var consensusMod modules.ConsensusModule
+	genesis := runtime.GetGenesis()
+
+	consM, err := consensus.Create(runtime)
+	consensusMod := consM.(modules.ConsensusModule)
```
Error handle before casting ^

andrewnguyen22 commented on consensus/consensus_tests/pacemaker_test.go:
```diff
@@ -133,7 +133,7 @@ func TestPacemakerCatchupSameStepDifferentRounds(t *testing.T) {
 		NumTxs:            0,
 		LastBlockHash:     "",
-		ProposerAddress:   leader.Address.Bytes(),
+		ProposerAddress:   leader.GetP2PAddress().Bytes(),
```
The `P2PAddress` should not be the address contained in the block header as it's just an ephemeral P2P identity. It should be the `Consensus Address`

deblasis replied:
This should fix it

https://github.com/pokt-network/pocket/pull/235/files#diff-c474ba0a5fb5b813e523913230483776eb1344705fb79db6d267ced0566c85f8R132-R141

andrewnguyen22 commented on consensus/consensus_tests/utils_test.go:
```diff
@@ -127,21 +133,20 @@ func CreateTestConsensusPocketNode(
-	pocketNode := &shared.Node{
-		Address: pk.Address(),
-	}
+	pocketNode := shared.NewWithAddress(pk.Address())
```
This function name is very non-descript. Please make it clearer what object is being returned from `shared` package

deblasis replied:
Agreed 

https://github.com/pokt-network/pocket/pull/235/files#diff-9f97d711d54398bf6257c266207519c11eb1d254c798d578786d0e94713cb164R151

andrewnguyen22 commented on consensus/consensus_tests/utils_test.go:
```diff
@@ -176,21 +181,21 @@ func StartAllTestPocketNodes(t *testing.T, pocketNodes IdToNodeMapping) {
 // define the interfaces used for debug/development. The latter will probably scale more but will
 // require more effort and pollute the source code with debugging information.
-func GetConsensusNodeState(node *shared.Node) typesCons.ConsensusNodeState {
+func GetConsensusNodeState(node modules.NodeModule) typesCons.ConsensusNodeState {
```
I don't think this change in naming convention made it clearer (I think the opposite)

deblasis replied:
I am not sure I understand what you mean. 

Is the fact that we are accepting the interface or the name of the interface that sounds odd?

andrewnguyen22 commented on consensus/leader_election/module.go:
```diff
@@ -16,13 +16,18 @@ type LeaderElectionModule interface {
 }
 
-func Create(_ *typesCons.ConsensusConfig, _ *typesCons.ConsensusGenesisState) (LeaderElectionModule, error) {
+func Create(runtime modules.Runtime) (modules.Module, error) {
```
This create wrapper seems like a clunkier design, considering the function signatures are the same.

Have we considered `new(consensusModule).Create(runtime)` instead of using a wrapper for each?

deblasis replied:
🤦‍♂️ you are totally right

https://github.com/pokt-network/pocket/pull/235/files#diff-eba95a8b904f8186683d2f9c106a3dc021f014918420e6a074d0a2b4e59c261dR66-R68

deblasis replied:
Without the wrapper we would have to expose all the structs and actually in this round of changes, I made all modules structs unexported

andrewnguyen22 commented on consensus/module.go:
```diff
@@ -65,47 +64,53 @@ type ConsensusModule struct {
+}
+
+func (*ConsensusModule) Create(runtime modules.Runtime) (modules.Module, error) {
+	var m *ConsensusModule
```
It doesn't make sense to me that you have a receiver variable that is ignored, but then the line below you declare another consensus module `var m *ConsensusModule`

deblasis replied:
🤦‍♂️ you are totally right

https://github.com/pokt-network/pocket/pull/235/files#diff-eba95a8b904f8186683d2f9c106a3dc021f014918420e6a074d0a2b4e59c261dR66-R68

andrewnguyen22 commented on consensus/module.go:
```diff
@@ -124,53 +129,20 @@ func Create(configPath, genesisPath string, useRandomPK bool) (modules.Consensus
 		utilityContext:    nil,
-		paceMaker:         paceMaker,
-		leaderElectionMod: leaderElectionMod,
+		paceMaker:         pacemakerMod,
```
You renamed the variable `pacemakerMod` but kept the structure field the legacy `pacemaker` let's be consistent with that

andrewnguyen22 commented on p2p/module_raintree_test.go:
```diff
@@ -15,12 +15,14 @@ import (
 	"testing"
```
```golang
for i := 0; i < numAddressessToBeAdded; i++ {
				newAddr, _ := crypto.GenerateAddress()
				network.AddPeerToAddrBook(&types.NetworkPeer{Address: newAddr})
			}
```

Catch the error in `newAddr, _ := crypto.GenerateAddress()`

deblasis replied:
Addressed here

https://github.com/pokt-network/pocket/pull/235/files#diff-2a2651c6ad9e1f7d34a394f137a8da69f8204d60e0d22ba7b304cb48d72e5d77R118

andrewnguyen22 commented on p2p/module_raintree_test.go:
```diff
@@ -15,12 +15,14 @@ import (
 	"github.com/golang/mock/gomock"
 	typesP2P "github.com/pokt-network/pocket/p2p/types"
 	mocksP2P "github.com/pokt-network/pocket/p2p/types/mocks"
+	typesPers "github.com/pokt-network/pocket/persistence/types"
```
This is a step backwards when it comes to encapsulation. You've introduced cross module imports. Please revert / fix to use `test_artifacts`

deblasis replied:
Not happening anymore 

https://github.com/pokt-network/pocket/pull/235/files#diff-a8351346d15469e6fb8938609c72b7b4387e5d77fd634fc64ed82327ad40fb5dR4-R25

andrewnguyen22 commented on persistence/types/converters.go:
```diff
@@ -0,0 +1,149 @@
+	}
+}
+
+func ToPersistenceActors(actors []modules.Actor) []*Actor {
```
Another case where cross module importing is introduced. (Used by p2p)

deblasis replied:
This doesn't happen anymore

andrewnguyen22 commented on telemetry/prometheus_module.go:
```diff
@@ -33,22 +34,31 @@ const (
+func (m *PrometheusTelemetryModule) Create(runtime modules.Runtime) (modules.Module, error) {
+	cfg := runtime.GetConfig()
+	if err := m.ValidateConfig(cfg); err != nil {
+		log.Fatalf("config validation failed: %v", err)
```
I don't really understand the pattern of not returning the error and logging fatal instead. (**Ditto for all module creation**)

deblasis replied:
Addressed in https://github.com/pokt-network/pocket/pull/235/commits/2a686db94104405ac0ff6b2320c740bbf32a11ae

Olshansk replied:
Need to update this one as well

andrewnguyen22 commented on runtime/runtime.go:
```diff
@@ -0,0 +1,116 @@
+	"strings"
+
+	"github.com/mitchellh/mapstructure"
+	"github.com/pokt-network/pocket/consensus/types"
```
Let's be consistent with import naming `consensus/types` to `consTypes`

deblasis replied:
Addressed in https://github.com/pokt-network/pocket/pull/235/commits/a8530d2ab589c248d53352a8df2b37d351b362ac

andrewnguyen22 commented on runtime/config.go:
```diff
@@ -0,0 +1,102 @@
+var _ modules.TelemetryConfig = &Config{}
+var _ modules.UtilityConfig = &Config{}
+
+type Config struct {
```
There is a very similar structure in `shared/modules/types.go` that should be consolidated.

andrewnguyen22 commented on runtime/config.go:
```diff
@@ -0,0 +1,102 @@
+var _ modules.TelemetryConfig = &Config{}
+var _ modules.UtilityConfig = &Config{}
+
+type Config struct {
```
I really prefer if this was an exported interface and not an exported structure

deblasis replied:
the renamed `runtimeConfig` now implements `modules.Config`.

runtime package exports the constructor and the `BaseConfig`

https://github.com/pokt-network/pocket/pull/235/commits/1924e0e7e3c6988d13cb0b4a2a4edc5c07c64f2b#diff-22e2eb4a42003f34d95bacbe4bf7dfa8d3ddd6ec3eb66c5ed611b5b6b72de26bR14-R23

andrewnguyen22 commented on shared/modules/consensus_module.go:
```diff
@@ -11,6 +11,9 @@ type ValidatorMap map[string]Actor // TODO (Drewsky) deprecate Validator map or
 
 type ConsensusModule interface {
 	Module
+	ConfigurableModule
```
Question: Are there other `modules` that are not configurable and genesis dependent?

I would like to see if we can move some of these interfaces to `Module`

deblasis replied:
I'd rather keep interfaces separated to be honest.

It's true that the majority of modules are `ConfigurableModule` but only a couple are `GenesisDependentModule`.

I'd like to keep `Module` to the bare minimum.

andrewnguyen22 commented on shared/modules/types.go:
```diff
@@ -15,6 +15,8 @@ type GenesisState struct {
 type BaseConfig struct {
 	RootDirectory string `json:"root_directory"`
 	PrivateKey    string `json:"private_key"` // TODO (pocket/issues/150) better architecture for key management (keybase, keyfiles, etc.)
+	ConfigPath    string `json:"config_path"`
```
These fields that are introduced are not used anywhere. What's the reasoning?

Also, we have a very similar structure under `runtime` package. What's the plan there?

deblasis replied:
Refactored mostly in https://github.com/pokt-network/pocket/pull/235/commits/1924e0e7e3c6988d13cb0b4a2a4edc5c07c64f2b

Relying more on abstractions

andrewnguyen22 commented on shared/modules/bus_module.go:
```diff
@@ -27,6 +25,6 @@ type Bus interface {
 	// Configuration
-	GetConfig() map[string]json.RawMessage
-	GetGenesis() map[string]json.RawMessage
+	GetConfig() Config
```
Why are these methods re-introduced?

deblasis replied:
Addressed in https://github.com/pokt-network/pocket/pull/235/commits/2837b5aff91973a67988c6d5db9ae893a6445437

andrewnguyen22 commented on utility/test/module_test.go:
```diff
@@ -63,7 +65,7 @@ func NewTestingUtilityContext(t *testing.T, height int64) utility.UtilityContext
 func newTestPersistenceModule(databaseUrl string) modules.PersistenceModule {
 	cfg := modules.Config{
-		Persistence: &test_artifacts.MockPersistenceConfig{
+		Persistence: &typesPers.PersistenceConfig{
```
Cross module import

deblasis replied:
Maybe addressed in https://github.com/pokt-network/pocket/pull/235/commits/1924e0e7e3c6988d13cb0b4a2a4edc5c07c64f2b#diff-e31d62c553295c0f69210e9d518a19d448a13f20bb057ad8d94b58a13bfcfa05L11

deblasis committed:
fix(Shared): bus creation via runtime

deblasis committed:
refactor(Shared): NewWithAddress -> NewNodeWithAddress

deblasis committed:
refactor(Shared): improved Create, making all modules unexported

deblasis committed:
refactor(Consensus): consistency in naming paceMaker

deblasis committed:
refactor(Shared): runtime -> runtimeCfg

deblasis committed:
refactor(Shared): code review feedback

deblasis committed:
Merge remote-tracking branch 'upstream/main' into issue/config_genesis_refactoring1

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>

deblasis committed:
fix(Shared): bugfixes

deblasis committed:
Merge remote-tracking branch 'upstream/main' into issue/config_genesis_refactoring1

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>

deblasis committed:
fix(Consensus): merge fix

deblasis committed:
test(Consensus): fixed tests (StoreBlock and InsertBlock mocks cfg)

deblasis committed:
fix(Shared): removed config and genesis from bus

deblasis committed:
style(Runtime): cleanup

deblasis committed:
style(Runtime): import alias

deblasis committed:
fix(Shared): module creation failures returns err

deblasis committed:
refactor(Persistence): reverted converters to shared

deblasis commented:
Clarification about the fact that cross-module imports were still present when I requested review.

I mistakenly thought that the issue (https://github.com/pokt-network/pocket/issues/239) mentioned in response to this https://github.com/pokt-network/pocket/pull/235#discussion_r975789014 was a form of de-scoping from the current issue and instead was to track via automation if we have them.

My bad, sorry guys. 

Also I have noticed that it should really be (removed quotes for exclude-dir):
```bash
grep --exclude-dir=persistence -r "github.com/pokt-network/pocket/persistence"
grep --exclude-dir=consensus -r "github.com/pokt-network/pocket/consensus"
...
```

deblasis committed:
feat(Tooling): check_cross_module_imports makefile target

deblasis committed:
style(Runtime): import names

Olshansk commented on Makefile:
```diff
@@ -383,3 +383,24 @@ gen_genesis_and_config:
 clear_genesis_and_config:
 	rm build/config/gen.*.json
+
+.PHONY: check_cross_module_imports
```
No-op.

In @okdas's check on linters, we have the ability to create custom linters in go: https://github.com/pokt-network/pocket/pull/190

I think we might be able to port this there.

deblasis committed:
Merge remote-tracking branch 'upstream/main' into issue/config_genesis_refactoring1

deblasis committed:
refactor(Shared): refactored clock to be sourced from runtimeMgr

deblasis committed:
fix(Makefile): excluding makefile!

deblasis committed:
feat(Tooling): improved check_cross_module_imports

deblasis commented:
I would like to know what's the plan of action with the current cross-module imports. 
Not sure if they are in scope for this PR since there is also other/older TECHDEBT in the mix.

I have updated/improved `make check_cross_module_imports` in https://github.com/pokt-network/pocket/pull/235/commits/72f6c90272ac9f901450fb077e06020c04c6fd98 to ignore stuff that's obviously dependent on these packages. 
Things like the `Makefile` itself, `shared`, `app`, `runtime`.

Result:

![image](https://user-images.githubusercontent.com/29378614/193247776-33de3b27-babf-488a-b3b6-b03ae31ad63a.png)

I am sure @okdas can make it even better of point me to further improvements 🙏

deblasis committed:
docs(Shared): Updated README and CHANGELOGs

deblasis committed:
docs(Runtime): missing items

deblasis committed:
test(Persistence): Tests: added missing mock configuration

Closes 262

deblasis committed:
Merge branch 'issue/262-persistence-failing-tests' into issue/config_genesis_refactoring1

deblasis committed:
fix(Tooling): execute tests sequentially fix (wrong syntax)

deblasis committed:
Merge branch 'issue/262-persistence-failing-tests' into issue/config_genesis_refactoring1

Olshansk commented:
@deblasis I'm going to do another look through this PR from start-to-finish and prioritize it for the beginning of this week.

> I would like to know what's the plan of action with the current cross-module imports.
> Not sure if they are in scope for this PR since there is also other/older TECHDEBT in the mix.

I believe this PR was a net improvement in terms of configuration management but we have cross-module imports both before and after.

If there are any "extremely offensive" cross-module imports, we should remove them here, but otherwise we should create a followup ticket and enable automation to prevent this once they're all gone.

@andrewnguyen22 After another round of reviews from both of us, does that sound like a reasonable plan?

Olshansk commented on consensus/types/consensus_genesis.go:
```diff
@@ -0,0 +1,18 @@
+	return ActorsToActorsInterface(x.GetValidators())
+}
+
+func ActorsToActorsInterface(a []*Validator) (actorI []modules.Actor) {
```
s/a/vals

Olshansk commented on consensus/types/converters.go:
```diff
@@ -0,0 +1,20 @@
+}
+
+func ToConsensusValidators(actors []modules.Actor) []*Validator {
+	r := make([]*Validator, 0)
+	for _, a := range actors {
+		r = append(r, actorToValidator(a))
+	}
+	return r
```
```suggestion
	vals = make([]*Validator, len(actors))
	for i, actor := range actors {
		vals[i] = actor
	}
	return
```

Olshansk commented on p2p/module.go:
```diff
@@ -28,7 +27,7 @@ const (
 type p2pModule struct {
 	bus       modules.Bus
-	p2pConfig modules.P2PConfig // TODO (olshansky): to remove this since it'll be available via the bus
+	p2pConfig *typesP2P.P2PConfig // TODO (olshansky): to remove this since it'll be available via the bus
```
In consensus, we renamed `consConfig` to `config` but get `p2pConfig` here. 

My preference is to keep explicit naming for the explicit struct (p2pConfig), but moreso to just be consistent between the two regardless of the approach.

Olshansk commented on persistence/types/converters.go:
```diff
@@ -0,0 +1,151 @@
+	}
+}
+
+func ToPersistenceActors(actors []modules.Actor) []*Actor {
+	r := make([]*Actor, 0)
+	for _, a := range actors {
+		r = append(r, toPersistenceActor(a))
+	}
+	return r
+}
```
I think if you search for `, 0)` in another couple of places, the same approach can be adopted.

Olshansk commented on utility/test/module_test.go:
```diff
@@ -67,58 +65,27 @@ func NewTestingUtilityContext(t *testing.T, height int64) utility.UtilityContext
+
+	err = persistenceMod.Start()
+	require.NoError(t, err)
+	return persistenceMod.(modules.PersistenceModule)
```
```suggestion

	return persistenceMod.(modules.PersistenceModule)
```

deblasis committed:
Merge remote-tracking branch 'upstream/main' into issue/config_genesis_refactoring1

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>

deblasis committed:
Merge remote-tracking branch 'upstream/main' into issue/config_genesis_refactoring1

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>

deblasis committed:
style(Makefile): check_cross_module_imports

deblasis committed:
Update app/client/main.go

Co-authored-by: Daniel Olshansky <olshansky.daniel@gmail.com>

deblasis committed:
refactor(Consensus): keys -> validatorKeys

deblasis committed:
Update consensus/consensus_tests/utils_test.go

Co-authored-by: Daniel Olshansky <olshansky.daniel@gmail.com>

deblasis commented on Makefile:
```diff
@@ -392,3 +392,24 @@ gen_genesis_and_config:
+
+.PHONY: check_cross_module_imports
+## Lists cross-module imports
+check_cross_module_imports:
```
I previously tried to do it with eval, I spent like 5 mins on it, I "dropped the ball" and moved on to something else.

Now with renewed motivation it worked immediately, the culprit is always in *escaping* quotes 🤣

deblasis committed:
Merge branch 'issue/config_genesis_refactoring1' of github.com:deblasis/pocket into issue/config_genesis_refactoring1

deblasis committed:
Update consensus/consensus_tests/utils_test.go

Co-authored-by: Daniel Olshansky <olshansky.daniel@gmail.com>

deblasis committed:
refactor(Consensu): NewNodeWithAddress -> NewNodeWithP2PAddress

deblasis committed:
fix(Makefile): revert silent

deblasis committed:
style(Shared): init var before return

deblasis committed:
Update shared/test_artifacts/generator.go

Co-authored-by: Daniel Olshansky <olshansky.daniel@gmail.com>

deblasis committed:
Update utility/test/module_test.go

Co-authored-by: Daniel Olshansky <olshansky.daniel@gmail.com>

deblasis committed:
Merge branch 'issue/config_genesis_refactoring1' of github.com:deblasis/pocket into issue/config_genesis_refactoring1

deblasis committed:
Update consensus/module.go

Co-authored-by: Daniel Olshansky <olshansky.daniel@gmail.com>

deblasis committed:
Update consensus/types/converters.go

Co-authored-by: Daniel Olshansky <olshansky.daniel@gmail.com>

deblasis commented on consensus/types/consensus_genesis.go:
```diff
@@ -0,0 +1,18 @@
+
+var _ modules.ConsensusGenesisState = &ConsensusGenesisState{}
+
+func (x *ConsensusGenesisState) GetVals() []modules.Actor {
```
We can't because there's already a `func (x *ConsensusGenesisState) GetValidators() []*Validator `

deblasis committed:
refactor(Consensus): s/ActorListToMap/ActorListToValidatorMap

deblasis committed:
Update consensus/types/converters.go

Co-authored-by: Daniel Olshansky <olshansky.daniel@gmail.com>

deblasis committed:
Merge branch 'issue/config_genesis_refactoring1' of github.com:deblasis/pocket into issue/config_genesis_refactoring1

deblasis committed:
fix(Consensus): converters fix

deblasis committed:
refactor(Shared): preallocations

deblasis commented on shared/README.md:
```diff
@@ -54,11 +55,11 @@ flowchart TD
-            B <-.-> Clock
         end
         subgraph Pocket's Core Modules
+            RMGR(RuntimeMgr)
```
I don't think it's relevant at this level, sounds like an implementation detail. 
It would make the diagram quite "busy" IMHO, also I could argue that `config` and `genesis` are probably more important than clock in the context of `runtimeMgr`

deblasis committed:
docs(Shared): nits

deblasis committed:
style(Persistence): cleanup

deblasis committed:
Update shared/modules/module.go

Co-authored-by: Daniel Olshansky <olshansky.daniel@gmail.com>

deblasis committed:
fix(Telemetry): error management

deblasis committed:
Update shared/modules/module.go

Co-authored-by: Daniel Olshansky <olshansky.daniel@gmail.com>

deblasis commented on shared/modules/types.go:
```diff
@@ -72,6 +73,7 @@ type ConsensusGenesisState interface {
 	GetGenesisTime() *timestamppb.Timestamp
 	GetChainId() string
 	GetMaxBlockBytes() uint64
+	GetVals() []Actor
```
Can't do. There's already a GetValidator

deblasis commented on shared/modules/node_module.go:
```diff
@@ -0,0 +1,8 @@
+package modules
```
The idea was something we could mock later on. since I see I have embedded `P2PAddressableModule` in there, I think I was thinking about P2P at the time of writing.

andrewnguyen22 replied:
Blocker:

We need comments here to describe this thing.

I agree with Olshansk

> I don't fully understand the logic behind this interface. Can you expand on the thought process? I think we might need to remove it.

deblasis commented on shared/node.go:
```diff
@@ -16,56 +17,69 @@ import (
 	MainModuleName = "main"
 )
 
 type Node struct {
```
https://github.com/pokt-network/pocket/pull/235#discussion_r986192819

deblasis committed:
Merge remote-tracking branch 'upstream/main' into issue/config_genesis_refactoring1

deblasis committed:
style(P2P): renamed p2pCfg

deblasis commented on p2p/module_raintree_test.go:
```diff
@@ -371,85 +367,75 @@ func prepareConnMock(t *testing.T, expectedNumNetworkReads, expectedNumNetworkWr
+	mockGenesisState := createMockGenesisState(t, valKeys)
+	for i := range mockRuntimeMgrs {
+		mockConfig := modulesMock.NewMockConfig(ctrl)
+		mockConfig.EXPECT().GetBaseConfig().Return(&runtime.BaseConfig{
```
I think the scope of this PR has been stretched already quite a bit, it's going to become harder and harder to maintain if we keep adding stuff :)

Nothing is overkill but time kills

Olshansk commented:
The only blocker from my end is fixing the unit tests. Some of it might be resolved after we merge in #282, so let's aim to get that in first.

<img width="413" alt="Screen Shot 2022-10-06 at 10 02 35 AM" src="https://user-images.githubusercontent.com/1892194/194375468-4679c3da-8f2c-44be-9163-2c3fc647d9ea.png">

----

I took another look through the PR. With the goal of unblocking downstream work. the 3 biggest action items remaining IMO are (see screenshots below):

1. Consider renaming `NodeModules`
2. Figuring out how to implement `ValidateConfig` while allowing injection of mocks
3. Renaming `GetVals` to `GetValidators` because `Vals` can be misinterpreted

@deblasis Can you create a "235 follow ups" to track it?

@andrewnguyen22 Assuming we merge in #282 and create the follow-up for that work, can we merge this in?

---

<img width="737" alt="Screen Shot 2022-10-06 at 10 00 49 AM" src="https://user-images.githubusercontent.com/1892194/194374730-e6e637a5-2a90-4890-91ac-730a02d576d9.png">
<img width="739" alt="Screen Shot 2022-10-06 at 10 00 35 AM" src="https://user-images.githubusercontent.com/1892194/194374737-e2476135-5850-4099-a52f-5a70fbe7b479.png">
<img width="734" alt="Screen Shot 2022-10-06 at 10 00 27 AM" src="https://user-images.githubusercontent.com/1892194/194374738-674026b9-d7d9-48b6-a064-fbd597ad6952.png">

andrewnguyen22 commented:
I will review it again @Olshansk

deblasis committed:
Merge remote-tracking branch 'upstream/main' into issue/config_genesis_refactoring1

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>

deblasis commented:
I have created the issues to track the open items as requested. They are not assigned to me and I put them in QoL, please change that as you see fit.

> The only blocker from my end is fixing the unit tests. Some of it might be resolved after we merge in #282, so let's aim to get that in first.

I have to point out that tests are not failing in this PR... what's failing is the CI because of the new linting: 

![image](https://user-images.githubusercontent.com/29378614/194411364-8de8bbf0-2a90-4a5a-b15a-d69d7a43ffed.png)

and also for the build / push part:

![image](https://user-images.githubusercontent.com/29378614/194411580-0fa65685-6fe1-4f00-a6f2-67c08b40df33.png)

Olshansk reviewed (approved):
Thanks for creating the follow up work.

👍  from my POV. Not perfect but I was to prevent the scope of the PR from increasing and it's already a big improvement over what we had.

Let's wait to hear if @andrewnguyen22 has any blockers though

andrewnguyen22 commented on consensus/module.go:
```diff
@@ -75,48 +73,53 @@ type ConsensusModule struct {
-		consCfg:     cfg,
-		consGenesis: genesis,
+		consCfg:     cfg.GetConsensusConfig().(*typesCons.ConsensusConfig),
+		consGenesis: genesis.GetConsensusGenesisState().(*typesCons.ConsensusGenesisState),
```
Non-Blocker: 

Change this to safe catch error:

```

casted, ok := cfg.GetConsensusConfig().(*typesCons.ConsensusConfig)
if !ok {
  // handle error
}
```

andrewnguyen22 replied:
Same with LeaderElectionModule casting below (and all other casting)

deblasis replied:
Maybe addressed with https://github.com/pokt-network/pocket/pull/235#discussion_r993986279

andrewnguyen22 commented on consensus/hotstuff_leader.go:
```diff
@@ -306,10 +315,11 @@ func (m *ConsensusModule) validatePartialSignature(msg *typesCons.HotstuffMessag
-//       Add proper tests and implementation once the mempool is implemented.
-func (m *ConsensusModule) tempIndexHotstuffMessage(msg *typesCons.HotstuffMessage) {
+//
+//	It doesn't actually work because SizeOf returns the size of the map pointer,
```
Non-Blocker: 

Revert, Goland doesn't recognize this as part of the TODO anymore... Maybe that was the intention

deblasis replied:
Done, VSCode formats files on save automatically, I need to configure it to mimic Goland I suppose. I'll look into it

andrewnguyen22 commented on consensus/module.go:
```diff
@@ -205,28 +175,48 @@ func (m *ConsensusModule) Start() error {
+}
+
+func (*consensusModule) ValidateGenesis(genesis modules.GenesisState) error {
+	// DISCUSS (team): we cannot cast if we want to use mocks and rely on interfaces
```
Exactly, see comment above

deblasis replied:
Maybe addressed with https://github.com/pokt-network/pocket/pull/235#discussion_r993986279

andrewnguyen22 commented on consensus/module.go:
```diff
@@ -75,48 +73,53 @@ type ConsensusModule struct {
 		privateKey:  privateKey.(cryptoPocket.Ed25519PrivateKey),
-		consCfg:     cfg,
-		consGenesis: genesis,
+		consCfg:     cfg.GetConsensusConfig().(*typesCons.ConsensusConfig),
```
Blocker:

We should not cast here because this prevents us from using mocks etc. 

We should use interfaces only

deblasis replied:
We are now using interfaces and I think I found a solution for the problem: "What if there are Consensus modules that have a Pacemaker and others that don't?" (and similar use-cases).

andrewnguyen22 commented on p2p/module.go:
```diff
@@ -27,8 +26,8 @@ const (
-	bus       modules.Bus
-	p2pConfig modules.P2PConfig // TODO (olshansky): to remove this since it'll be available via the bus
+	bus    modules.Bus
+	p2pCfg *typesP2P.P2PConfig // TODO (olshansky): to remove this since it'll be available via the bus
```
Blocker:

Another example of the casting, this should be `modules.P2PConfig`

deblasis replied:
Maybe addressed with https://github.com/pokt-network/pocket/pull/235#discussion_r993986279

andrewnguyen22 commented on p2p/module.go:
```diff
@@ -119,12 +100,12 @@ func (m *p2pModule) Start() error {
 	}
 
-	if m.p2pConfig.GetUseRainTree() {
+	if m.p2pCfg.UseRainTree {
```
Blocker:

Revert once using interface

deblasis replied:
Maybe addressed with https://github.com/pokt-network/pocket/pull/235#discussion_r993986279

andrewnguyen22 commented on p2p/module_raintree_test.go:
```diff
@@ -367,85 +363,75 @@ func prepareConnMock(t *testing.T, expectedNumNetworkReads, expectedNumNetworkWr
-
-func createConfigs(t *testing.T, numValidators int) (configs []modules.Config, genesisState modules.GenesisState) {
-	configs = make([]modules.Config, numValidators)
+func createMockRuntimeMgrs(t *testing.T, numValidators int) []modules.RuntimeMgr {
```
Blocker:

Please add comment around any `Mock` creation, see the other mocks in this file for example around expected behaviors

andrewnguyen22 commented on p2p/module_raintree_test.go:
```diff
@@ -367,85 +363,75 @@ func prepareConnMock(t *testing.T, expectedNumNetworkReads, expectedNumNetworkWr
 }
 
-func createGenesisState(t *testing.T, valKeys []cryptoPocket.PrivateKey) modules.GenesisState {
+func createMockGenesisState(t *testing.T, valKeys []cryptoPocket.PrivateKey) modules.GenesisState {
```
Blocker:

Please add comment around any Mock creation, see the other mocks in this file for example around expected behaviors

andrewnguyen22 commented on runtime/manager.go:
```diff
@@ -0,0 +1,174 @@
+	return mgr
+}
+
+func NewManagerFromReaders(configReader, genesisReader io.Reader, options ...func(*Manager)) *Manager {
```
Blocker:

No usage function with no comments or tests. Please add one of the two

deblasis replied:
We could potentially remove this. I have used during development and forgot to clean after myself

andrewnguyen22 commented on shared/modules/module.go:
```diff
@@ -21,6 +24,21 @@ type InterruptableModule interface {
 	GetModuleName() string
-	InitConfig(pathToConfigJSON string) (IConfig, error)
-	InitGenesis(pathToGenesisJSON string) (IGenesis, error)
+	Create(runtime RuntimeMgr) (Module, error)
```
Non-Blocker:

The granularity here is getting a bit sticky for me. Consider mental load vs. verbosity

deblasis replied:
I see. If it sounds like over-optimization or premature optimization let's consider the fact that soon we'll have modules that are very simple and only implement a subset of the Module interface. Logging for example, which is coming soon. Keybase as well I guess.

Maybe my .Net background surfaces here. I am used to very small and targeted interfaces.

I am open to suggestions.

andrewnguyen22 commented on shared/test_artifacts/generator.go:
```diff
@@ -34,28 +40,31 @@ var (
+	serviceNodes, snPrivateKeys := NewActors(types.ActorType_ServiceNode, numServiceNodes)
+	fish, fishPrivateKeys := NewActors(types.ActorType_Fisherman, numFisherman)
+
+	genesisState := runtime.NewGenesis(
```
Non-Blocker:

Is there no way to use `runtime` for this?

deblasis replied:
If you mean to move test_artifacts under runtime. I have done it.

andrewnguyen22 commented on shared/test_artifacts/gov.go:
```diff
@@ -1,9 +1,11 @@
 package test_artifacts
```
Non-Blocker:

Can we move this to the runtime package?

deblasis replied:
Done

andrewnguyen22 commented on telemetry/noop_module.go:
```diff
@@ -62,6 +59,14 @@ func (m *NoopTelemetryModule) GetBus() modules.Bus {
 }
 
+func (*NoopTelemetryModule) ValidateConfig(cfg modules.Config) error {
+	// DISCUSS (team): we cannot cast if we want to use mocks and rely on interfaces
```
Agree, see applicable comment above

deblasis replied:
Maybe addressed with https://github.com/pokt-network/pocket/pull/235#discussion_r993986279

andrewnguyen22 commented on utility/module.go:
```diff
@@ -1,78 +1,75 @@
 }
+
+func (*utilityModule) ValidateConfig(cfg modules.Config) error {
+	// DISCUSS (team): we cannot cast if we want to use mocks and rely on interfaces
```
Agree see applicable comment above

deblasis replied:
Maybe addressed with https://github.com/pokt-network/pocket/pull/235#discussion_r993986279

andrewnguyen22 reviewed (changes requested):
Pretty much there, a few blockers and a few discussion items. 

Resolve conflicts and blockers and I'll approve

deblasis committed:
Merge remote-tracking branch 'upstream/main' into issue/config_genesis_refactoring1

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>

deblasis committed:
fix(Tooling): reverted a line change introduced in #282

It was causing test failures because of inconsistent Postgresql state

deblasis committed:
refactor(Shared): modules use interfaces for config and genesis

deblasis committed:
refactor(Shared): using interfaces vs structs

deblasis committed:
style(Shared): formatting

Goland is the offender... weird format on save rules... investigate

deblasis committed:
style(Consensus): cleanup

deblasis committed:
docs(Runtime): NewManagerFromReaders comment

deblasis committed:
refactor(Shared): shared.test_artifacts -> runtime.test_artifacts

deblasis committed:
fix(Persistence): leftover casts

deblasis committed:
refactor(Shared): removed NodeModule

deblasis commented:
@andrewnguyen22 thank you for your time and feedback reviewing this  :) 

I have addressed all the comments that I have marked with a 👍 and left comments otherwise. There are a couple of questions from me but nothing major, I guess.

Regarding `NodeModule` I think that since both of you flagged it, its short life had to come to an end. I have axed it and replaced with the struct. Perhaps I was over-abstracting.

andrewnguyen22 reviewed (approved):
Great work @deblasis
